### PR TITLE
[SYCL][ESIMD] Introduce predicates for lsc_block_store/load

### PR DESCRIPTION
**Pull request by fineg74**

### CODE REVIEW

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 1]
```diff
@@ -450,10 +451,11 @@ template <typename T, uint8_t NElts = 1,
           typename AccessorTy>
 __ESIMD_API std::enable_if_t<!std::is_pointer<AccessorTy>::value,
                              __ESIMD_NS::simd<T, NElts>>
-lsc_block_load(AccessorTy acc, uint32_t offset) {
+lsc_block_load(AccessorTy acc, uint32_t offset,
+               __ESIMD_NS::simd_mask<1> pred = 1) {
 #ifdef __ESIMD_FORCE_STATELESS_MEM
   return lsc_block_load<T, NElts, DS, L1H, L3H>(
-      __ESIMD_DNS::accessorToPointer<T>(acc, offset));
+      __ESIMD_DNS::accessorToPointer<T>(acc, offset, pred));
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:458]: `pred` should apparently be an argument of `lsc_block_load`, not `accessorToPointer`.
This should have been caught by the https://github.com/intel/llvm-test-suite/pull/1194 test, right? Do you have idea why it wasn't? Probably, the test should also compile/run with forced stateless memory access

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 2]
```diff
@@ -442,6 +442,7 @@ lsc_gather(AccessorTy acc, __ESIMD_NS::simd<uint32_t, N> offsets,
 /// @tparam AccessorTy is the \ref sycl::accessor type.
 /// @param acc is the SYCL accessor.
 /// @param offset is the zero-based offset in bytes.
+/// @param pred is predicate.
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:445]: Please add description of the effect of the predicate

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 3]
```diff
@@ -861,17 +863,19 @@ lsc_scatter(AccessorTy acc, __ESIMD_NS::simd<uint32_t, N> offsets,
...
 lsc_block_store(AccessorTy acc, uint32_t offset,
-                __ESIMD_NS::simd<T, NElts> vals) {
+                __ESIMD_NS::simd<T, NElts> vals,
+                __ESIMD_NS::simd_mask<1> pred = 1) {
 #ifdef __ESIMD_FORCE_STATELESS_MEM
   lsc_block_store<T, NElts, DS, L1H>(
-      __ESIMD_DNS::accessorToPointer<T>(acc, offset), vals);
+      __ESIMD_DNS::accessorToPointer<T>(acc, offset), vals, pred);
 #else
   detail::check_lsc_vector_size<NElts>();
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:880]: Please make sure new functionality also works on ESIMD emulator, as `lsc` accesses work on emulator. @dongkyunahn-intel can help with logistics.

**Review by kbobrovs — CHANGES_REQUESTED** [review]
Please address comments.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 3 above)
Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:883]: I locally verified if two samples from https://github.com/intel/llvm-test-suite/pull/1194 are working fine with this PR. While `lsc_predicate.cpp` passed, `lsc_predicate_stateless.cpp` failed during compilation.

Building dpcpp toolchain supporting ESIMD_EMULATOR and running ESIMD kernels on emulator requires some environment setup. Let me know if you need help in locally verifying your implementation with ESIMD_EMULATOR.

Comment (dongkyunahn-intel) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:877]: After `__ESIMD_FORCE_STATELESS_MEM` fix, both tests are passing.

```
$ SYCL/ESIMD/lsc/Output/lsc_predicate.cpp.tmp.out
Device name: ESIMD_EMULATOR
USM lsc predicate test passed
USM lsc predicate test passed
USM lsc predicate test passed
Accessor lsc predicate test passed
Accessor lsc predicate test passed
Accessor lsc predicate test passed
$ SYCL/ESIMD/lsc/Output/lsc_predicate_stateless.cpp.tmp.out
Device name: ESIMD_EMULATOR
Accessor lsc predicate test passed
Accessor lsc predicate test passed
Accessor lsc predicate test passed
```

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 4]
```diff
@@ -427,6 +472,46 @@ lsc_gather(AccessorTy acc, __ESIMD_NS::simd<uint32_t, N> offsets,
...
+/// Collects elements located at specified address and returns them
+/// as a single \ref simd object.
+///
+/// @tparam T is element type.
+/// @tparam NElts is the number of elements to load per address.
+/// @tparam DS is the data size.
+/// @tparam L1H is L1 cache hint.
+/// @tparam L3H is L3 cache hint.
+/// @param p is the base pointer.
+/// @param pred is predicate to enable the operation
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:488]:
```suggestion
/// @param pred operation predicate. Zero means operation is skipped entirely, non-zero - operation is performed.
```

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 5]
```diff
@@ -442,6 +527,8 @@ lsc_gather(AccessorTy acc, __ESIMD_NS::simd<uint32_t, N> offsets,
 /// @tparam AccessorTy is the \ref sycl::accessor type.
 /// @param acc is the SYCL accessor.
 /// @param offset is the zero-based offset in bytes.
+/// @param pred is predicate to enable the operation
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:530]: here please also expand description a bit telling which values mean what

**Review by kbobrovs — DISMISSED** [review]

**Review by kbobrovs — APPROVED** [review]

**Comment by kbobrovs** [comment]
@fineg74, please file internal ticket (if it does not exist yet) for CUDA testing failures.

**Pull request merged**